xilem_web/src/attribute.rs

```diff
@@ -1,55 +1,302 @@
+                            self.updated_attributes.remove(name);
+                            set_attribute(element, name, &value.serialize());
+                            // element.set_attribute(name, &value.serialize());
+                        }
+                    }
+                    AttributeModifier::EndMarker(_) => (),
+                }
+            }
+            debug_assert!(self.updated_attributes.is_empty());
+        }
+        self.build_finished = true;
+    }
+}
+
+impl WithAttributes for Attributes {
+    fn set_attribute(&mut self, name: CowStr, value: Option<AttributeValue>) {
+        let new_modifier = if let Some(value) = value {
+            AttributeModifier::Set(name.clone(), value)
+        } else {
+            AttributeModifier::Remove(name.clone())
+        };
+
+        if let Some(modifier) = self.attribute_modifiers.get_mut(self.idx) {
+            if modifier != &new_modifier {
+                if let AttributeModifier::Remove(previous_name)
+                | AttributeModifier::Set(previous_name, _) = modifier
+                {
+                    if &name != previous_name {
+                        self.updated_attributes.insert(previous_name.clone(), ());
+                    }
+                }
+                self.updated_attributes.insert(name, ());
+                *modifier = new_modifier;
+            }
+            // else remove it out of updated_attributes? (because previous attributes are overwritten) not sure if worth it because potentially worse perf
+        } else {
+            self.updated_attributes.insert(name, ());
+            self.attribute_modifiers.push(new_modifier);
+        }
+        self.idx += 1;
+    }
+
+    fn start_attribute_modifier(&mut self) {
+        if self.build_finished {
+            if self.idx == 0 {
+                self.start_idx = 0;
+            } else {
+                let AttributeModifier::EndMarker(start_idx) =
+                    self.attribute_modifiers[self.idx - 1]
+                else {
+                    unreachable!("this should not happen, as either `start_attribute_modifier` happens first, or follows an end_attribute_modifier")
+                };
+                self.idx = start_idx;
+                self.start_idx = start_idx;
+            }
+        }
+    }
+
+    fn end_attribute_modifier(&mut self) {
+        match self.attribute_modifiers.get_mut(self.idx) {
+            Some(AttributeModifier::EndMarker(prev_start_idx))
+                if *prev_start_idx == self.start_idx => {} // class modifier hasn't changed
+            Some(modifier) => {
+                *modifier = AttributeModifier::EndMarker(self.start_idx);
+            }
+            None => {
+                self.attribute_modifiers
+                    .push(AttributeModifier::EndMarker(self.start_idx));
+            }
+        }
+        self.idx += 1;
+        self.start_idx = self.idx;
+    }
+}
```

Comment:
This is one of the more interesting parts of this PR, a modifier based system (`Styles`, `Classes` and this `Attributes`), which is designed in such a way, that `Memoize` or `Arc` doesn't offer weird surprises (as well as just overwriting attributes, when they're previously set, which would have problems, when just directly setting attributes on the DOM elements).

As well as `OneOf` views being able to intersect on the elements and their props in a strong/statically-typed way. For example when there is `OneOf2::A(html::div(()))` and `OneOf2::B(mathml::mrow(())`, this `OneOf2` view fulfills the contract of the `Element` trait, but not of the `HtmlElement`, because the trait bound `AsRef<HtmlElement>` is not satisfied for `mathml::mrow`, so e.g. `HtmlElement::style` is not possible to use (currently at least, because MathML *should* be able to support style, this needs extension in wasm_bindgen I think.), but all the methods of `Element` are possible on `mathml::mrow` as well.

The modifier system works in such a way, that it has stateful `start` and `end` methods, which add markers into a `Vec` for the modifier type on its props to know where the modifier view has started/ended.
Additional modifiers for this property (e.g. `AttributeModifier::Set` or `AttributeModifier::Remove`) are then pushed into this `Vec` until the next modifier starts. 
All of this happens in either `View::build` as well as `View::rebuild`, and these attributes are added on every reconciliation (i.e. `View::rebuild`), so that it can detect changes (so all these attributes should be added always in the same order, to avoid unnecessary updates).
On each reconciliation pass, it is checked, whether the attribute at the same index has changed, and if so, the relevant props or attributes are marked dirty, such that they're correctly updated when applying the props to the actual DOM element.
This should be robust to underlying element changes in e.g. a `OneOf` which would build the new view, while being in a `View::rebuild` of a composing modifier view, as the actual modifiers have to be applied on the up-traversal and only the `end` method actually tells via an `EndMarker` where the modifier has started (which is read by the `start` method on the down traversal, to jump to the next modifier).

All the possible accumulated changes are then applied when `SuperElement::upcast`ing the element, which usually happens on a `ViewSequence` boundary.
These modifiers are described via traits such as (currently), `WithAttributes`, `WithClasses` and `WithStyles` and are also working with the intersection of `OneOf` types.